pgmpy/tests/test_base/test_AncestralBase.py

```diff
@@ -0,0 +1,235 @@
+        self.assertEqual(self.graph["B"]["C"]["marks"], {"B": ">", "C": "-"})
+        self.assertEqual(self.graph["C"]["D"]["marks"], {"C": "o", "D": "o"})
+
+    def test_add_edge_same_node_error(self):
+        """Test that adding edge with same source and target raises error."""
+        with self.assertRaises(ValueError):
+            self.graph.add_edge("A", "A", "-", ">")
+
+    def test_add_edge_invalid_marks_error(self):
+        """Test that invalid marks raise ValueError."""
+        with self.assertRaises(ValueError):
+            self.graph.add_edge("A", "B", "x", ">")
+        with self.assertRaises(ValueError):
+            self.graph.add_edge("A", "B", "-", "y")
+        with self.assertRaises(ValueError):
+            self.graph.add_edge("A", "B", "z", "w")
+
+    def test_add_edges_from(self):
+        """Test adding multiple edges from list."""
+        edges = [("A", "B", "-", ">"), ("B", "C", ">", "-"), ("A", "C", "o", "o")]
+        self.graph.add_edges_from(edges)
+
+        self.assertEqual(len(self.graph.edges), 3)
+        self.assertEqual(len(self.graph.nodes), 3)
+        self.assertEqual(self.graph["A"]["B"]["marks"], {"A": "-", "B": ">"})
+        self.assertEqual(self.graph["B"]["C"]["marks"], {"B": ">", "C": "-"})
+        self.assertEqual(self.graph["A"]["C"]["marks"], {"A": "o", "C": "o"})
+
+    def test_get_neighbors_basic(self):
+        """Test getting neighbors without constraints."""
+        self.graph.add_edge("A", "B", "-", ">")
+        self.graph.add_edge("A", "C", ">", "-")
+
+        neighbors_a = self.graph.get_neighbors("A")
+        self.assertEqual(neighbors_a, {"B", "C"})
+
+        neighbors_b = self.graph.get_neighbors("B")
+        self.assertEqual(neighbors_b, {"A"})
+
+    def test_get_neighbors_nonexistent_node(self):
+        """Test getting neighbors for non-existent node."""
+        neighbors = self.graph.get_neighbors("Z")
+        self.assertEqual(neighbors, set())
+
+    def test_get_parents(self):
+        """Test getting parent nodes."""
+        self.graph.add_edge("A", "B", "-", ">")
+        self.graph.add_edge("C", "B", "-", ">")
+        self.graph.add_edge("B", "D", "-", ">")
+
+        self.graph.add_edge("X", "B", "o", "o")
+        self.graph.add_edge("Y", "B", ">", ">")
+        self.graph.add_edge("Z", "B", ">", "-")
+
+        parents_b = self.graph.get_parents("B")
+        self.assertEqual(parents_b, {"A", "C", "Y"})
+
+        parents_d = self.graph.get_parents("D")
+        self.assertEqual(parents_d, {"B"})
+
+        parents_a = self.graph.get_parents("A")
+        self.assertEqual(parents_a, set())
+
+    def test_get_children(self):
+        """Test getting child nodes."""
+        self.graph.add_edge("A", "B", "-", ">")
+        self.graph.add_edge("A", "C", "-", ">")
+        self.graph.add_edge("B", "D", "-", ">")
+
+        self.graph.add_edge("A", "X", "o", "o")
+        self.graph.add_edge("A", "Z", ">", "-")
```

Comment:
Would suggest adding a bidirected edge in the graph here as well. Will help in catching issues like the ones we saw above.